### PR TITLE
Add ConfigSecret element to the AlertmanagerSpec.

### PR DESCRIPTION
**Pull request by AngusDavis**
Issue #2003 describes an issue with Kustomize naming of secrets and the
name used for discovering the alertmanager configuration secret. In a
nutshell, a secret created with kustomize will include the hash of the
secret values as a suffix of the object name which is unexpected by
alertmanager. Instead; when an alertmanager is created, its pods will
wait for a secret of the appropriate name to be created, when in reality
they will will not be.

This change adds a ConfigSecret element to the AlertmanagerSpec that
allows the name of the secret to be injected into the stateful set.

This change is strongly influenced by PR #2024; by @lswith.

Addresses #2003.

### CODE REVIEW

File: Documentation/api.md [hunk 1]
```diff
@@ -127,6 +127,7 @@ AlertmanagerSpec is a specification of the desired behavior of the Alertmanager
 | imagePullSecrets | An optional list of references to secrets in the same namespace to use for pulling prometheus and alertmanager images from registries see http://kubernetes.io/docs/user-guide/images#specifying-imagepullsecrets-on-a-pod | [][v1.LocalObjectReference](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#localobjectreference-v1-core) | false |
 | secrets | Secrets is a list of Secrets in the same namespace as the Alertmanager object, which shall be mounted into the Alertmanager Pods. The Secrets are mounted into /etc/alertmanager/secrets/<secret-name>. | []string | false |
 | configMaps | ConfigMaps is a list of ConfigMaps in the same namespace as the Alertmanager object, which shall be mounted into the Alertmanager Pods. The ConfigMaps are mounted into /etc/alertmanager/configmaps/<configmap-name>. | []string | false |
+| configSecret | ConfigSecret is the name of a Kubernetes Secret in the same namespace as the AlertManager object, which contains configuration for this Alertmaanger instance. Defaults to 'alertmanager-<alertmanager-name>' The secret is mounted into /etc/alertmanager/config. | string | false |
```
Comment (brancz) [inline Documentation/api.md]: Two typos in “Alertmanager” (the m is not capitalized)

Comment (AngusDavis) [reply Documentation/api.md]: Thanks, I've changed one instance of AlertManager in this comment to Alertmanager to be consistent with the rest of the docs in this section. I've also fixed an instance of 'Alertmaanger' to be spelled Alertmanager.

**Comment by brancz** [comment]
Sorry looks like this needs a rebase and then re-generate.

**Comment by AngusDavis** [comment]
> Sorry looks like this needs a rebase and then re-generate.

Thanks, wasn't certain of whether to wait until the end to rebase or rebase as we go. I've rebased, keeping the CR commit in the middle to show changes.

**Comment by AngusDavis** [comment]
Looks like known flake #2823 ; amending the commit msg & forcing a push to re-trigger travis.

**Comment by brancz** [comment]
Thanks!

**Comment by AngusDavis** [comment]
Thanks for the review, @brancz!

**Pull request merged**